File: src/Features/CSharp/Portable/ValidateFormatString/ValidateFormatStringDiagnosticAnalyzer.cs
// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

using System;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Diagnostics;
using Microsoft.CodeAnalysis.ValidateFormatString;

namespace Microsoft.CodeAnalysis.CSharp.ValidateFormatString
{
    [DiagnosticAnalyzer(LanguageNames.CSharp)]
    internal sealed class ValidateFormatStringDiagnosticAnalyzer : AbstractValidateFormatStringDiagnosticAnalyzer
    {
        public override void Initialize(AnalysisContext context)
        {
            context.RegisterSyntaxNodeAction(AnalyzeNode, SyntaxKind.InvocationExpression);
        }

        private static void AnalyzeNode(SyntaxNodeAnalysisContext context)
        {
            var invocationExpr = (InvocationExpressionSyntax)context.Node;
            if (invocationExpr.ArgumentList?.Arguments == null)
            {
                return;
            }

            SimpleNameSyntax name;
            // When calling string.Format(...), Format will be of type MemberAccessExpressionSyntax 
            if (invocationExpr.Expression is MemberAccessExpressionSyntax memberAccessExpr)
            {
                if (memberAccessExpr.Name?.ToString() != "Format")
                {
                    return;
                }

                name = memberAccessExpr.Name;
            }
            // When using static System.String and calling Format(...), Format will be of type IdentifierNameSyntax
            else if (invocationExpr.Expression is IdentifierNameSyntax identifierNameSyntax)
            {
                if (identifierNameSyntax.ToString() != "Format")
                {
                    return;
                }

                name = identifierNameSyntax;
            }
            else
            {
                return;
            }

            var arguments = invocationExpr.ArgumentList.Arguments;
            var numberOfArguments = arguments.Count;

            var symbolInfo = context.SemanticModel.GetSymbolInfo(name, context.CancellationToken);
           
            if (!TryGetFormatMethod(context, numberOfArguments, symbolInfo, out var method))
            {
                return;
            }

            var hasIFormatProvider =
                method.Parameters[0].Type.ToString().Equals(typeof(IFormatProvider).FullName);

            if (!TryGetFormatStringAndLocation(
                arguments,
                hasIFormatProvider,
                out var formatString,
                out var formatStringLocation))
            {
                return;
            }

            ValidateAndReportDiagnostic(context, hasIFormatProvider, arguments.Count, formatString, formatStringLocation);
        }

        private static bool TryGetFormatStringAndLocation(
            SeparatedSyntaxList<ArgumentSyntax> arguments, 
            bool hasIFormatProvider, 
            out string formatString, 
            out Location formatStringLocation)
        {
            formatString = null;
            formatStringLocation = null;

            var formatArgumentSyntax = GetFormatStringArgument(arguments, hasIFormatProvider);
            if (formatArgumentSyntax == null)
            {
                return false;
            }

            if (!formatArgumentSyntax.Expression.IsKind(SyntaxKind.StringLiteralExpression))
            {
                return false;
            }

            var formatLiteralExpressionSyntax = (LiteralExpressionSyntax)formatArgumentSyntax.Expression;
            formatString = formatLiteralExpressionSyntax.Token.ValueText;
            formatStringLocation = formatArgumentSyntax.GetLocation();

            return true;
        }

        private static ArgumentSyntax GetFormatStringArgument(SeparatedSyntaxList<ArgumentSyntax> arguments, bool hasIFormatProvider)
        {
            foreach (var argument in arguments)
            {
                if (argument.NameColon != null && !argument.NameColon.IsMissing && argument.NameColon.Name.Identifier.ValueText.Equals("format"))
                {
                    return argument;
                }
            }

            // If using positional arguments, the format string will be the first or second
            // argument depending on whether there Is an IFormatProvider argument.
            return (hasIFormatProvider ? arguments[1] : arguments[0]);
        }
    }
}

File: src/Features/VisualBasic/Portable/ValidateFormatString/ValidateFormatStringDiagnosticAnalyzer.vb
' Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
Imports Microsoft.CodeAnalysis.Diagnostics
Imports Microsoft.CodeAnalysis.ValidateFormatString

Namespace Microsoft.CodeAnalysis.VisualBasic.ValidateFormatString
    <DiagnosticAnalyzer(LanguageNames.VisualBasic)>
    Friend Class ValidateFormatStringDiagnosticAnalyzer
        Inherits AbstractValidateFormatStringDiagnosticAnalyzer

        Public Overrides Sub Initialize(context As AnalysisContext)
            context.RegisterSyntaxNodeAction(AddressOf AnalyzeNode, SyntaxKind.InvocationExpression)
        End Sub

        Private Sub AnalyzeNode(context As SyntaxNodeAnalysisContext)
            Dim invocationExpr = CType(context.Node, InvocationExpressionSyntax)
            If (invocationExpr.ArgumentList?.Arguments Is Nothing) Then
                Return
            End If

            Dim name As SimpleNameSyntax
            ' When calling string.Format(...), Format will be of type MemberAccessExpressionSyntax 
            Dim memberAccessExpr = TryCast(invocationExpr.Expression, MemberAccessExpressionSyntax)
            ' When using static System.String and calling Format(...), Format will be of type IdentifierNameSyntax
            Dim identifierNameSyntax = TryCast(invocationExpr.Expression, IdentifierNameSyntax)
            If (memberAccessExpr IsNot Nothing AndAlso memberAccessExpr.Name IsNot Nothing AndAlso memberAccessExpr.Name.ToString() = "Format") Then
                name = memberAccessExpr.Name
            ElseIf (identifierNameSyntax IsNot Nothing AndAlso identifierNameSyntax.ToString() = "Format") Then
                name = identifierNameSyntax
            Else
                Return
            End If


            Dim arguments = invocationExpr.ArgumentList.Arguments
            Dim numberOfArguments = arguments.Count

            Dim symbolInfo = context.SemanticModel.GetSymbolInfo(name, context.CancellationToken)

            Dim method As IMethodSymbol = Nothing

            If (Not TryGetFormatMethod(context, numberOfArguments, symbolInfo, method)) Then
                Return
            End If

            Dim hasIFormatProvider = (method.Parameters(0).Type.ToString() = GetType(IFormatProvider).ToString())

            Dim formatString As String = Nothing
            Dim formatStringLocation As Location = Nothing

            If (Not TryGetFormatStringAndLocation(arguments, hasIFormatProvider, formatString, formatStringLocation)) Then
                Return
            End If

            ValidateAndReportDiagnostic(context, hasIFormatProvider, arguments.Count, formatString, formatStringLocation)
        End Sub

        Private Function TryGetFormatStringAndLocation(arguments As SeparatedSyntaxList(Of ArgumentSyntax), hasIFormatProvider As Boolean, ByRef formatString As String, ByRef formatStringLocation As Location) As Boolean
            formatString = Nothing
            formatStringLocation = Nothing

            Dim formatArgumentSyntax = GetFormatStringArgument(arguments, hasIFormatProvider)

            If (formatArgumentSyntax Is Nothing) Then
                Return False
            End If

            If (Not formatArgumentSyntax.IsKind(SyntaxKind.SimpleArgument)) Then
                Return False
            End If

            Dim simpleArguementSyntax = TryCast(formatArgumentSyntax, SimpleArgumentSyntax)
            If (simpleArguementSyntax Is Nothing) Then
                Return False
            End If

            If (simpleArguementSyntax.Expression Is Nothing) Then
                Return False
            End If

            formatString = simpleArguementSyntax.Expression.ToString
            formatStringLocation = formatArgumentSyntax.GetLocation

            Return True
        End Function

        Private Function GetFormatStringArgument(arguments As SeparatedSyntaxList(Of ArgumentSyntax), hasIFormatProvider As Boolean) As ArgumentSyntax
            For Each argument In arguments
                If (argument.IsNamed) Then
                    Dim simpleArgumentSyntax = TryCast(argument, SimpleArgumentSyntax)
                    If simpleArgumentSyntax IsNot Nothing AndAlso simpleArgumentSyntax.NameColonEquals.Name.Identifier.ValueText.Equals("format") Then
                        Return argument
                    End If
                End If
            Next
            ' If using positional arguments, the format string will be the first or second
            ' argument depending on whether there Is an IFormatProvider argument.
            Return If(hasIFormatProvider, arguments(1), arguments(0))
        End Function
    End Class
End Namespace
